core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java

```diff
@@ -96,39 +82,64 @@ public static boolean ingestNode(Settings settings) {
 
     public static final List<DiscoveryNode> EMPTY_LIST = Collections.emptyList();
 
-    private String nodeName = "";
-    private String nodeId;
-    private String hostName;
-    private String hostAddress;
-    private TransportAddress address;
-    private ImmutableOpenMap<String, String> attributes;
-    private Version version = Version.CURRENT;
+    private final String nodeName;
+    private final String nodeId;
+    private final String hostName;
+    private final String hostAddress;
+    private final TransportAddress address;
+    private final ImmutableOpenMap<String, String> attributes;
+    private final Version version;
+    private final Set<Role> roles;
 
-    DiscoveryNode() {
+    /**
+     * Creates a new {@link DiscoveryNode} by reading from the stream provided as argument
+     * @param in the stream
+     * @throws IOException if there is an error while reading from the stream
+     */
+    public DiscoveryNode(StreamInput in) throws IOException {
+        this.nodeName = in.readString().intern();
+        this.nodeId = in.readString().intern();
+        this.hostName = in.readString().intern();
+        this.hostAddress = in.readString().intern();
+        this.address = TransportAddressSerializers.addressFromStream(in);
+        int size = in.readVInt();
+        ImmutableOpenMap.Builder<String, String> attributesBuilder = ImmutableOpenMap.builder(size);
+        for (int i = 0; i < size; i++) {
+            attributesBuilder.put(in.readString().intern(), in.readString().intern());
+        }
+        this.attributes = attributesBuilder.build();
+        int rolesSize = in.readVInt();
+        this.roles = new HashSet<>(rolesSize);
```

Comment:
Also I think this should be an `EnumSet` and maybe just read using a switch statement?